modules/fx/arrow-fx/src/test/kotlin/arrow/fx/IOTest.kt

```diff
@@ -171,34 +173,57 @@ class IOTest : UnitSpec() {
       }
     }
 
-    "should not catch exceptions within run block with unsafeRunAsync" {
+    "should return exceptions within main block with unsafeRunAsyncCancellable" {
+      val exception = MyException()
+      val ioa = IO<Int> { throw exception }
+      ioa.unsafeRunAsyncCancellable { either ->
+        either.fold({ it shouldBe exception }, { fail("") })
+      }
+    }
+
+    "should rethrow exceptions within run block with unsafeRunAsync" {
       try {
         val exception = MyException()
         val ioa = IO<Int> { throw exception }
         ioa.unsafeRunAsync { either ->
-          either.fold({ throw exception }, { fail("") })
+          either.fold({ throw it }, { fail("") })
         }
         fail("Should rethrow the exception")
       } catch (myException: MyException) {
         // Success
       } catch (throwable: Throwable) {
-        fail("Should only throw MyException")
+        fail("Should only throw MyException but was $throwable")
       }
     }
 
-    "should complete when running a pure value with runAsync" {
-      val expected = 0
-      just(expected).runAsync { either ->
-        either.fold({ fail("") }, { IO { it shouldBe expected } })
+    "should rethrow exceptions within run block with unsafeRunAsyncCancelable" {
+      try {
+        val exception = MyException()
+        val ioa = IO<Int> { throw exception }
+        ioa.unsafeRunAsyncCancellable { either ->
+          either.fold({ throw it }, { fail("") })
+        }
+        fail("Should rethrow the exception")
+      } catch (myException: MyException) {
+        // Success
+      } catch (throwable: Throwable) {
+        fail("Should only throw MyException but was $throwable")
       }
     }
 
-    "should complete when running a return value with runAsync" {
```

Comment:
@pakoito I'm not sure this tests are really doing anything, if I change the `shouldBe` part to expect something else nothing's gonna fail here, is that intended?

Reply:
Does it fail, or is the failure just swallowed?

In any case, don't leave dead code in. Remove it and we can return it later.

Reply:
it's swallowed, yep

Reply:
I am still in favor of figuring out a solution for `runAsync` in `0.10.x` and deprecating it so we can remove it.
Operator makes no sense and it throws everyone off :D

In Cats-effect, it's being replaced by the following. The signature was meant to build edges of the world for `F`, and now `Effect` can be derived through `MonadIO` for `F ~> IO` which means you can build a polymorphic edge of the world. We can achieve the same with `suspend`, but for `ConcurrentEffect` (cancelation) we need support in the language. I would be in favor of doing both tho.

```
trait Effect[F[_]] {
   def to[G[_], A](fa: F[A]): G[A] =
}
```

Reply:
-> #2010